### ISSUE TITLE
FLUIDNC scraper compatibility.

### ISSUE THREAD

**Issue by brakthehun** [arkypita/LaserGRBL#1885]
FluidNC reports motion control a little differently through $axes request and $spindle request.
Current issue is the homing cycle / max move speed / limits / spindle power / laser mode uses a $$ request which FluidNC doesn't respond to the same as GRBL. Only returns $10

**Comment by arkypita** [comment]
Is this a FluidNC issue or a LaserGRBL issue?

Can you give a more detailed example of the messages sent by FluidNC?
If possible, can you log the whole communication (LaserGRBL, menu "?", "activate extended log"

**Comment by brakthehun** [comment]
![2022-08-16 21_47_08-Window](https://user-images.githubusercontent.com/16564592/185036970-0fa35cf1-dd49-4fa1-8d0e-884ca7594322.png)
![2022-08-16 21_44_29-Window](https://user-images.githubusercontent.com/16564592/185036973-b90a1161-13ea-4fc8-a172-e511a84ce3c9.png)
![2022-08-16 21_45_01-Window](https://user-images.githubusercontent.com/16564592/185036974-51ba65c6-039c-4bfe-a8bc-cd80c3c2ef64.png)
![2022-08-16 21_46_14-Window](https://user-images.githubusercontent.com/16564592/185036975-c803e362-8dda-45e0-87ee-0abeb3a7b48d.png)

**Comment by brakthehun** [comment]
[comlog.txt](https://github.com/arkypita/LaserGRBL/files/9356426/comlog.txt)

**Comment by arkypita** [comment]
LaserGRBL is compatible with GRBL, and with firmware capable of behaving like GRBL.

Based on what [I read here](http://wiki.fluidnc.com/en/features/grbl_compatibility) FluidNC it has basic compatibility, and as far as I can see from comlog.txt it seems to me that work quite well.

Feature like $axes and $spindle and other things specific to this firmware are not part of the GRBL specification and therefore are not supported in LaserGRBL.

**Comment by arkypita** [comment]
Is there anything I did not understand in your request? Can you rephrase it?

**Comment by brakthehun** [comment]
I would like the homing button to show up and the speed max to be set.
I guess I'm asking LaserGRBL to support FluidNC

**Comment by brakthehun** [comment]
When scraping,, The max speed is set by :
$100
$101
In FluidNC:
$axes/X/max_rate_per_min:
$axes/Y/max_rate_per_min:

Laser power max is stored in:
$laser/speedmap:
instead of
$30

**Comment by buster3** [comment]
I implemented the $30 command in FluidNC. Still, the Focus Toggle in LaserGRBL is not working for me. Any hints? Thanks

![image](https://user-images.githubusercontent.com/6793480/194752168-23e04744-369c-4a10-b40c-9a43368b2c8c.png)

**Comment by arkypita** [comment]
What version of LaserGRBL are you using?  Look like an old version that does not support formula in custom button.

**Comment by buster3** [comment]
I use the latest v4.9.4

**Comment by buster3** [comment]
Any hints here? Should the formula be evaluated by fluidnc or by laserGRBL?

**Comment by arkypita** [comment]
By LaserGRBL, but LaserGRBL should be able to read $30 value from controller to be able to evaluate formula.
If this is not possible, just edit custom button and put a fixed S value based on your needs.

**Comment by brakthehun** [comment]
That's the issue with the laserGRBL scraper. 
With fluidNC you can query $30 but it doesn't return any value.
You would have to query:
$laser/speed_map
And it would return one or two mins and a max. The max would be the $30 value.

**Comment by buster3** [comment]
> That's the issue with the laserGRBL scraper. With fluidNC you can query $30 but it doesn't return any value. You would have to query: $laser/speed_map And it would return one or two mins and a max. The max would be the $30 value.

His does not hold any more for the last fluidNC release. See https://github.com/bdring/FluidNC/issues/648

**Comment by buster3** [comment]
> That's the issue with the laserGRBL scraper. With fluidNC you can query $30 but it doesn't return any value. You would have to query: $laser/speed_map And it would return one or two mins and a max. The max would be the $30 value.

His does not hold any more for the last fluidNC release. See https://github.com/bdring/FluidNC/issues/648

**Comment by buster3** [comment]
Hello, i revisit this issue. Can anybody give me a hint where the '$$' command is implemented. Can I simply add $30 as output? Thanks

**Comment by buster3** [comment]
I revisited this issue. Can anybody confirm that LaserGRBL uses $$ and expects a output for $30? How should the output format for $$ look like? Thanks

**Comment by arkypita** [comment]
LaserGRBL send $$ than expect to receive the "ok" response (acknowledge to $$).
After ok received, all the lines the macine send is queued into a list (until a timeout since the last message received) then this list is parsed to extract config parameters.

a config parameter is something like "$" at the beginning of a line, followed by numbers, by any spaces, by "=" and then everything. this regular expression is used: "^[$](\d+)\s*=(.*)"
